### PR TITLE
Ensure that no other RRsets exist alongside CNAME

### PR DESCRIPTION
**Pull request by peterthomassen**
This PR does not enforce the database-level constraint yet, as it requires installing a Postgres extension for which we currently cannot provide a migration, as Postgres superuser privileges are required. However, starting Postgres 13 (going to be released this fall), the pertinent extension is considered trusted, and the Django database user will be able to install it. So, once we upgraded to Postgres 13, let's add migrations equivalent to the following:

```sql
CREATE EXTENSION btree_gist; -- this is what did not work before Postgres 13

ALTER TABLE desecapi_rrset ADD CONSTRAINT cname_exclusivity EXCLUDE USING gist (
    domain_id WITH =,
    subname WITH =,
    int4(type = 'CNAME') WITH <> 
);
```

For corresponding Django support, see https://docs.djangoproject.com/en/3.1/ref/contrib/postgres/operations/#btreegistextension and https://docs.djangoproject.com/en/3.1/ref/contrib/postgres/constraints/#django.contrib.postgres.constraints.ExclusionConstraint.

### CODE REVIEW

**Comment by peterthomassen** [comment]
CNAME exclusivity is now implemented the same way as the usual uniqueness constraint is validated at the serializer level. There also is code at the model (I extended `RRset.validate_unique()`).

I had the solution at the model first, but ran into issues with the fact that it throws a Django `ValidationError`, not a DRF one. One has to catch and convert it, but it's hard to assign the error details to the correct serializer field. I looked how we do it for regular uniqueness constraints, and saw that it is done at the serializer only, so I followed that example.

This means that:

- The code in the model is dead (it's not called anywhere). Shall we keep it, in case someone calls `RRset.full_clean(validate_unique=True)`, or `RRset.validate_unique()` directly? We'd need to add tests and maintain it.
- Once the database-level exclusion constraint is there (have to wait for Postgres 13, see above), I can't see the benefit why we should have it at the model, so I would propose to drop it.

Another open issue is: when sending two RRsets for the same subname in one bulk request, one of which has type CNAME, then the new check does not catch the exclusivity violation, because it's based on a database query, but the database has neither RRset at that time. (Currently, such requests hit pdns, which rejects the operation, and we return 500.) There is no test for this now, but I verified it manually. Ways forward:

- Accept it for now, and wait for the Postgres exclusion constraint. Then, this would give a database integrity error, and we could return 409 conflict (as is introduced by the first commit in this PR). However, this is not a nice kind of error (very unspecific).
- Improve detection of this situation in the serializer, specifically in `RRsetListSerializer.to_internal_value()` which currently detects `(subname, type)` duplicates in bulk requests. The current algorithm generates an index of `(subname, type)`, and then looks up duplicates. We'd have to extend this with a second index of `(subname, type == 'CNAME')` to then find RRsets which match in `subname` but not in `type == 'CNAME'`. This is considerable complexity, and it's the same thing that has to be done 1.) at the database level, 2.) at the model (if we want that), 3.) at the serializer (as a validator like the uniqueness one, comparing existing RRsets), and 4.) at the serializer to compare things within a bulk request. It sounds wrong to do that. -- The Postgres constraint would catch it all, at the expense of a nice message. (On the other hand, the exception tells us the name of the database constraint that failed, so a hint can be returned to the client.)

Thoughts?

**Comment by peterthomassen** [comment]
To complicate things further, an e2e test is failing now. The test has a pre-existing AAAA record and sends two RRsets in a bulk request, namely a) the AAAA RRset, but with `records = []` so that it gets deleted, and b) a CNAME RRset with the same subname. This situation is legal and functions as an atomic "replace something with a CNAME" operation.

Unfortunately, the validator at the serializer rejects it, because the exclusivity check for the new CNAME RRset is done at a time when the AAAA RRset still exists in the database. The validator would have to know that it is going to be deleted. Should we extend it like that, or is it overengineered? -- I suppose the same is true for replacing a CNAME with something else.

Another approach would be to first run all delete operations (without committing the transaction) before running the exclusivity validation. I believe this runs counter to DRF's internal structure. ...

**Comment by nils-wisiol** [comment]
Let's invest some time into translating database integrity errors into nice error messages. This is something we should do anyhow, as for some types of requests, the api cannot make sure that validation passes (think of massively parallel requests). Then, only after dbapi accepted changes, we contact nslord to update the DNS.

**Comment by peterthomassen** [comment]
Ready for review! -- Based on #450 which should be reviewed first.

**Comment by peterthomassen** [comment]
Upgrade instructions:

```shell
cd /usr/local/src/desec-stack/
docker-compose stop api celery
docker-compose exec dbapi pg_dumpall > outputfile  # ~2s
docker-compose stop dbapi
cp -a /var/lib/docker/volumes/desec*_dbapi_postgres postgres-volume.bak
docker-compose rm -f dbapi
docker volume rm desec-stack_dbapi_postgres  # hyphen may not be required
git pull && docker-compose pull
docker-compose up -d dbapi
# wait a few seconds for dbapi init to finish
cat outputfile | docker-compose exec -T dbapi psql -d postgres  # ~3s
docker-compose up -d
# If all went well:
rm -f outputfile postgres-volume.bak/
```

File: api/desecapi/serializers.py [hunk 1]
```diff
@@ -366,22 +377,56 @@ def to_internal_value(self, data):
...
+                            for position in conflict_positions:
+                                types_by_position.setdefault(position, []).append(type_)
+                        # Sort by position, None at the end
+                        types_by_position = dict(sorted(types_by_position.items(), key=lambda x: (x[0] is None, x)))
+                        db_conflicts = types_by_position.pop(None, None)
+                        if db_conflicts: types_by_position['database'] = db_conflicts
+                        for position, types in types_by_position.items():
+                            types_by_position[position] = ', '.join(types)
+                        types_by_position = [f'{position} ({types})' for position, types in types_by_position.items()]
+                        types_by_position = ', '.join(types_by_position)
```
Comment (nils-wisiol) [inline api/desecapi/serializers.py:422]: maybe use [`pprint.pformat`](https://docs.python.org/3/library/pprint.html#pprint.pformat) to save some lines here

File: api/desecapi/tests/test_rrsets_bulk.py [hunk 2]
```diff
@@ -142,6 +142,16 @@ def test_bulk_post_missing_fields(self):
...
 
+    def test_bulk_patch_cname_exclusivity(self):
+        response = self.client.bulk_patch_rr_sets(
+            domain_name=self.my_rr_set_domain.name,
+            payload=[
+                {'subname': 'test', 'type': 'A', 'ttl': 3600, 'records': ['1.2.3.4']},
+                {'subname': 'test', 'type': 'CNAME', 'ttl': 3600, 'records': ['example.com.']},
+            ]
+        )
+        self.assertResponse(response, status.HTTP_400_BAD_REQUEST)
```
Comment (nils-wisiol) [inline api/desecapi/tests/test_rrsets_bulk.py:153]: also assert error msg

File: api/desecapi/validators.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        value = self.exclusion_condition[1]
+        source = serializer.fields[field_name].source
+        if serializer.instance is not None:
+            if source not in attrs:
+                attrs[source] = getattr(serializer.instance, source)
+        exclusion_method = qs_exclude if attrs[source] == value else qs_filter
+        return exclusion_method(qs, **{field_name: value})
+
+    def __call__(self, attrs, serializer, *args, **kwargs):
+        # Ignore validation if the many flag is set: RRsetListSerializer will take care of the constraint
```
Comment (nils-wisiol) [inline api/desecapi/validators.py:38]: move to class doc?

File: api/desecapi/validators.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+        # Ignore validation if any field is None
+        checked_values = [
+            value for field, value in attrs.items() if field in self.fields
+        ]
+        if None not in checked_values and qs_exists(queryset):
+            types = queryset.values_list('type', flat=True)
+            types = ', '.join(types)
+            message = self.message.format(types=types)
+            raise ValidationError(message, code='unique')
```
Comment (nils-wisiol) [inline api/desecapi/validators.py:55]: `code` wrong?

**Review by nils-wisiol — COMMENTED** [review]
🚀

File: api/desecapi/exception_handlers.py [hunk 5]
```diff
@@ -55,6 +61,9 @@ def _503():
         metrics.get('desecapi_database_unavailable').inc()
         return _503()
 
+    if isinstance(exc, IntegrityError):
+        return _409(exc)
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/exception_handlers.py]: done. Took another approach to allow for inheritance (not exact type match)

File: api/desecapi/models.py [hunk 6]
```diff
@@ -453,6 +453,7 @@ def create(self, contents=None, **kwargs):
         rrset = super().create(**kwargs)
         for content in contents or []:
             RR.objects.create(rrset=rrset, content=content)
+
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/models.py:459]: done

File: api/desecapi/serializers.py [hunk 7]
```diff
@@ -17,7 +16,6 @@
 
 from api import settings
 from desecapi import crypto, metrics, models
-from desecapi.exceptions import ConcurrencyException
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/serializers.py:20]: no, it's used in `DynDNS12Update.get()`

File: api/desecapi/tests/test_rrsets.py [hunk 8]
```diff
@@ -282,11 +282,16 @@ def test_create_my_rr_sets_subname_too_many_dots(self):
...
 
-
     def test_create_my_rr_sets_empty_payload(self):
         response = self.client.post_rr_set(self.my_empty_domain.name)
         self.assertContains(response, 'No data provided', status_code=status.HTTP_400_BAD_REQUEST)
 
+    def test_create_my_rr_sets_cname_two_records(self):
+        data = {'subname': 'sub', 'records': ['example.com.', 'example.org.'], 'ttl': 3600, 'type': 'CNAME'}
+        response = self.client.post_rr_set(self.my_domain.name, **data)
+        print(response.data)
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/tests/test_rrsets.py]: done

File: api/desecapi/models.py [hunk 9]
```diff
@@ -519,8 +519,11 @@ def clean_records(self, records_presentation_format):
         rdtype = rdatatype.from_text(self.type)
         errors = []
 
-        if self.type == 'CNAME' and len(records_presentation_format) > 1:
-            errors.append('RRset of type CNAME cannot have multiple records.')
+        if self.type == 'CNAME':
+            if self.subname == '':
+                errors.append('CNAME RRset cannot have empty subname.')
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/models.py:536]: I don't think so

File: api/desecapi/serializers.py [hunk 10]
```diff
@@ -360,23 +360,25 @@ def to_internal_value(self, data):
...
         except TypeError:  # in case self.instance is None (as during POST)
             known_instances = {}
-        indices_by_key = {}
+        indices = {}
         for idx, item in enumerate(data):
             # Validate item type before using anything from it
             if not isinstance(item, dict):
                 self.fail('invalid', datatype=type(item).__name__)
-            items = indices_by_key.setdefault(self._key(item), set())
+            key = self._key(item)
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/serializers.py]: > `s, t = self._key(item)` to improve readability? `indices[s][t]` looks nice

done

> maybe even better to hide the dict creation altogether

still considering, will get back later

> I see this being reused in `collapsed_indices` -- maybe it's good to create a class for this data structure

`collapsed_indices` is only used once with `[s][t]` and once with `[s]`. I don't see how another data structure would make it simpler.

File: api/desecapi/serializers.py [hunk 11]
```diff
@@ -366,22 +377,56 @@ def to_internal_value(self, data):
             if not isinstance(item, dict):
                 self.fail('invalid', datatype=type(item).__name__)
             key = self._key(item)
-            items = indices.setdefault(key[0], {}).setdefault(key[1], set())
+            if key[0] not in indices:
+                types = self.child.domain.rrset_set.filter(subname=key[0]).values_list('type', flat=True)
+                indices[key[0]] = {type_: {None} for type_ in types}
```
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/serializers.py]: done

File: api/desecapi/serializers.py (hunk 1 above)
Comment (peterthomassen) [reply api/desecapi/serializers.py:422]: doesn't really work, will have extra apostrophes etc.

File: api/desecapi/tests/test_rrsets_bulk.py (hunk 2 above)
Comment (peterthomassen) [reply api/desecapi/tests/test_rrsets_bulk.py:153]: done

File: api/desecapi/validators.py (hunk 3 above)
Comment (peterthomassen) [reply api/desecapi/validators.py:38]: done

File: api/desecapi/validators.py (hunk 4 above)
Comment (peterthomassen) [reply api/desecapi/validators.py:55]: done

File: api/desecapi/serializers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (peterthomassen) [reply api/desecapi/serializers.py]: > maybe even better to hide the dict creation altogether

as discussed, left as is but moved dict reshuffling for error message creation to a separate method

**Comment by peterthomassen** [comment]
I took everything into account; ready for review

**Review by nils-wisiol — APPROVED** [review]
lgtm

**Pull request merged**